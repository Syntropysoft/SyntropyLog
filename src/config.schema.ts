/**
 * FILE: src/config.schema.ts
 * DESCRIPTION: Defines the Zod validation schemas for the entire library's configuration.
 * These schemas are the single source of truth for the configuration's structure and types.
 */

import { z } from 'zod';
import { Transport } from './logger/transports/Transport';
import { IHttpClientAdapter } from './http/adapters/adapter.types';
import { IBrokerAdapter } from './brokers/adapter.types';
import { MaskingStrategy } from './masking/MaskingEngine';

/**
 * @description Schema for logger-specific options, including serialization and transports.
 * @private
 */
const loggerOptionsSchema = z
  .object({
    name: z.string().optional(),
    level: z
      .enum(['fatal', 'error', 'warn', 'info', 'debug', 'trace', 'silent'])
      .optional(),
    serviceName: z.string().optional(),
    /**
     * An array of transport instances to be used by the logger.
     */
    transports: z.array(z.instanceof(Transport)).optional(),

    /**
     * A dictionary of custom serializer functions. The key is the field
     * to look for in the log object, and the value is the function that transforms it.
     */
    serializers: z
      .record(z.string(), z.function().args(z.any()).returns(z.string()))
      .optional(),
    /**
     * The maximum time in milliseconds a custom serializer can run before being timed out.
     * @default 50
     */
    serializerTimeoutMs: z.number().int().positive().default(50),

    /** Configuration for pretty printing logs in development. */
    prettyPrint: z
      .object({
        enabled: z.boolean().optional().default(false),
      })
      .optional(),
  })
  .optional();

/**
 * @description Reusable schema for retry options, commonly used in client configurations.
 * @private
 */
const retryOptionsSchema = z
  .object({
    maxRetries: z.number().int().positive().optional(),
    retryDelay: z.number().int().positive().optional(),
  })
  .optional();

/**
 * @description Schema for a single Redis instance, using a discriminated union for different connection modes.
 */
export const redisInstanceConfigSchema = z.discriminatedUnion('mode', [
  z.object({
    mode: z.literal('single'),
    instanceName: z.string(),
    url: z.string().url(),
    retryOptions: retryOptionsSchema,
    // --- NEW: Granular Logging Configuration for Redis ---
    logging: z
      .object({
        /** Level for successful commands. @default 'debug' */
        onSuccess: z.enum(['trace', 'debug', 'info']).default('debug'),
        /** Level for failed commands. @default 'error' */
        onError: z.enum(['warn', 'error', 'fatal']).default('error'),
        /** Whether to log command parameters. @default true */
        logCommandValues: z.boolean().default(true),
        /** Whether to log the return value of commands. @default false */
        logReturnValue: z.boolean().default(false),
      })
      .optional(),
  }),
  // Apply the same 'logging' object structure to 'sentinel' and 'cluster' modes
  z.object({
    mode: z.literal('sentinel'),
    instanceName: z.string(),
    name: z.string(),
    sentinels: z.array(z.object({ host: z.string(), port: z.number() })),
    sentinelPassword: z.string().optional(),
    retryOptions: retryOptionsSchema,
    logging: z
      .object({
        onSuccess: z.enum(['trace', 'debug', 'info']).default('debug'),
        onError: z.enum(['warn', 'error', 'fatal']).default('error'),
        logCommandValues: z.boolean().default(true),
        logReturnValue: z.boolean().default(false),
      })
      .optional(),
  }),
  z.object({
    mode: z.literal('cluster'),
    instanceName: z.string(),
    rootNodes: z.array(z.object({ host: z.string(), port: z.number() })),
    logging: z
      .object({
        /** Level for successful commands. @default 'debug' */
        onSuccess: z.enum(['trace', 'debug', 'info']).default('debug'),
        /** Level for failed commands. @default 'error' */
        onError: z.enum(['warn', 'error', 'fatal']).default('error'),
        /** Whether to log command parameters. @default true */
        logCommandValues: z.boolean().default(true),
        /** Whether to log the return value of commands. @default false */
        logReturnValue: z.boolean().default(false),
      })
      .optional(),
  }),
]);

/**
 * @description Schema for the main Redis configuration block, containing all Redis instances.
 */
export const redisConfigSchema = z
  .object({
    /** An array of Redis instance configurations. */
    instances: z.array(redisInstanceConfigSchema),
    /** The name of the default Redis instance to use when no name is provided to `getInstance()`. */
    default: z.string().optional(),
  })
  .optional();

/**
 * @description Schema for a single HTTP client instance.
 */
export const httpInstanceConfigSchema = z.object({
  instanceName: z.string(),
  adapter: z.custom<IHttpClientAdapter>((val) => {
    return (
      typeof val === 'object' &&
      val !== null &&
      'request' in val &&
      typeof (val as any).request === 'function'
    );
  }, "The provided adapter is invalid. It must be an object with a 'request' method."),
  isDefault: z.boolean().optional(),
  propagate: z.array(z.string()).optional(),
  propagateFullContext: z.boolean().optional(),
  logging: z
    .object({
      onSuccess: z.enum(['trace', 'debug', 'info']).default('info'),
      onError: z.enum(['warn', 'error', 'fatal']).default('error'),
      logSuccessBody: z.boolean().default(false),
      logSuccessHeaders: z.boolean().default(false),
      onRequest: z.enum(['trace', 'debug', 'info']).default('info'),
      logRequestBody: z.boolean().default(false),
      logRequestHeaders: z.boolean().default(false),
    })
    .partial()
    .optional(),
});

/**
 * @description Schema for the main HTTP configuration block.
 */
export const httpConfigSchema = z
  .object({
    /** An array of HTTP client instance configurations. */
    instances: z.array(httpInstanceConfigSchema),
    /** The name of the default HTTP client instance to use when no name is provided to `getInstance()`. */
    default: z.string().optional(),
  })
  .optional();

/**
 * @description Schema for the main data masking configuration block.
 */
const maskingConfigSchema = z
  .object({
    /** Array of masking rules with patterns and strategies. */
    rules: z.array(z.object({
      /** Regex pattern to match field names */
      pattern: z.union([z.string(), z.instanceof(RegExp)]),
      /** Masking strategy to apply */
      strategy: z.nativeEnum(MaskingStrategy),
      /** Whether to preserve original length */
      preserveLength: z.boolean().optional(),
      /** Character to use for masking */
      maskChar: z.string().optional(),
      /** Custom masking function (for CUSTOM strategy) */
      customMask: z.function().args(z.string()).returns(z.string()).optional(),
    })).optional(),
    /** Default mask character */
    maskChar: z.string().optional(),
    /** Whether to preserve original length by default */
    preserveLength: z.boolean().optional(),
    /** Enable default rules for common data types */
    enableDefaultRules: z.boolean().optional(),
  })
  .optional();

/**
 * @description Schema for a single message broker client instance.
 * It validates that a valid `IBrokerAdapter` is provided.
 * @private
 */
export const brokerInstanceConfigSchema = z.object({
  instanceName: z.string(),
  adapter: z.custom<IBrokerAdapter>((val) => {
    return (
      typeof val === 'object' &&
      val !== null &&
      typeof (val as any).publish === 'function' &&
      typeof (val as any).subscribe === 'function'
    );
  }, 'The provided broker adapter is invalid.'),
  /**
   * An array of context keys to propagate as message headers/properties.
   * To propagate all keys, provide an array with a single wildcard: `['*']`.
   * If not provided, only `correlationId` and `transactionId` are propagated by default.
   */
  propagate: z.array(z.string()).optional(),

  /**
   * @deprecated Use `propagate` instead.
   * If true, propagates the entire asynchronous context map as headers.
   * If false (default), only propagates `correlationId` and `transactionId`.
   */
  propagateFullContext: z.boolean().optional(),
  isDefault: z.boolean().optional(),
});

/**
 * @description Schema for the main message broker configuration block.
 */
export const brokerConfigSchema = z
  .object({
    /** An array of broker client instance configurations. */
    instances: z.array(brokerInstanceConfigSchema),
    /** The name of the default broker instance to use when no name is provided to `getInstance()`. */
    default: z.string().optional(),
  })
  .optional();

/**
 * @description Schema for the declarative logging matrix.
 * It controls which context properties are included in the final log output based on the log level.
 * @private
 */
const loggingMatrixSchema = z
  .object({
    /** An array of context keys to include in logs by default. Can be overridden by level-specific rules. */
    default: z.array(z.string()).optional(),
    /** An array of context keys to include for 'trace' level logs. Use `['*']` to include all context properties. */
    trace: z.array(z.string()).optional(),
    /** An array of context keys to include for 'debug' level logs. Use `['*']` to include all context properties. */
    debug: z.array(z.string()).optional(),
    /** An array of context keys to include for 'info' level logs. Use `['*']` to include all context properties. */
    info: z.array(z.string()).optional(),
    /** An array of context keys to include for 'warn' level logs. Use `['*']` to include all context properties. */
    warn: z.array(z.string()).optional(),
    /** An array of context keys to include for 'error' level logs. Use `['*']` to include all context properties. */
    error: z.array(z.string()).optional(),
    /** An array of context keys to include for 'fatal' level logs. Use `['*']` to include all context properties. */
    fatal: z.array(z.string()).optional(),
  })
  .optional();

/**
 * @description The main schema for the entire SyntropyLog configuration.
 * This is the single source of truth for validating the user's configuration object.
 */
export const syntropyLogConfigSchema = z.object({
  /** Logger-specific configuration. */
  logger: loggerOptionsSchema,

  /** Declarative matrix to control context data in logs. */
  loggingMatrix: loggingMatrixSchema,

  /** Redis client configuration. */
  redis: redisConfigSchema,
  /** HTTP client configuration. */
  http: httpConfigSchema,
  /** Message broker client configuration. */
  brokers: brokerConfigSchema,

  /** Centralized data masking configuration. */
  masking: maskingConfigSchema,

  /** Context propagation configuration. */
  context: z
    .object({
      /** The HTTP header name to use for the correlation ID. @default 'x-correlation-id' */
      correlationIdHeader: z.string().optional(),
      /** The HTTP header name to use for the external transaction/trace ID. @default 'x-trace-id' */
      transactionIdHeader: z.string().optional(),
    })
    .optional(),

  /**
   * The maximum time in milliseconds to wait for a graceful shutdown before timing out.
   * @default 5000
   */
  shutdownTimeout: z
    .number({
      description: 'The maximum time in ms to wait for a graceful shutdown.',
    })
    .int()
    .positive()
    .optional(),

  /** Configuration for the `syntropylog doctor` CLI tool. */
  doctor: z
    .object({
      /** An array of rule IDs to disable during a diagnostic run. */
      disableRules: z.array(z.string()).optional(),
    })
    .optional(),
});
