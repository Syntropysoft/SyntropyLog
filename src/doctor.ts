/*
=============================================================================
ARCHIVO 8: src/doctor.ts (NUEVO - API PÃšBLICA)
-----------------------------------------------------------------------------
DESCRIPTION (en-US):
This file serves as the public entry point for `syntropy/doctor`. It exports
the building blocks users will need to create their own custom audit manifests
and rules, such as `coreRules` and the necessary types.
=============================================================================
*/
export { coreRules } from './cli/checks';
export type { DiagnosticRule, CheckResult } from './cli/checks';