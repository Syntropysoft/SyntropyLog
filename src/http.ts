export type { InstrumentedFetch, InstrumentedHttpClient } from './http/types';
