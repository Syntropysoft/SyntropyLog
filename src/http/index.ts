/**
 * @file src/http/index.ts
 * DESCRIPTION:
 * This file is the public entry point for all components related to HTTP instrumentation.
 * It defines the public API that framework users will consume to create and use adapters.
 */

// 1. EXPORT THE CONTRACT INTERFACES
// This is the most important part. It allows ANYONE to create their own
// adapter by simply implementing these interfaces.
export type {
  IHttpClientAdapter,
  AdapterHttpRequest,
  AdapterHttpResponse,
  AdapterHttpError,
} from './adapters/adapter.types';

// 2. EXPORT THE ADAPTERS WE PROVIDE
// As a convenience for users, we export the adapters that we maintain
// ourselves (like the one for Axios). This way, they don't have to write
// it themselves if they use a common library.
export { AxiosAdapter } from './adapters/AxiosAdapter';

// 3. (OPTIONAL) EXPORT THE INSTRUMENTER
// Exporting the instrumenter class type can be useful for advanced
// TypeScript users who want explicit type references.
export type { InstrumentedHttpClient } from './InstrumentedHttpClient';
