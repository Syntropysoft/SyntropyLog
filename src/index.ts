/**
 * FILE: src/index.ts
 * DESCRIPTION: The main entry point for the SyntropyLog framework.
 * It exports the singleton instance of the framework and key types for users.
 */

export { syntropyLog } from './SyntropyLog';

// --- Core Types for Configuration and Usage ---
export type { SyntropyLogConfig } from './config';
export type { ILogger } from './logger/ILogger';
export type { IContextManager } from './context';

// --- Transport Classes and Interfaces for Custom Configuration ---
export { Transport } from './logger/transports/Transport';
export { ConsoleTransport } from './logger/transports/ConsoleTransport';
export { PrettyConsoleTransport } from './logger/transports/PrettyConsoleTransport';
export { CompactConsoleTransport } from './logger/transports/CompactConsoleTransport';
export { ClassicConsoleTransport } from './logger/transports/ClassicConsoleTransport'; // <-- ADDED
export { SpyTransport } from './logger/transports/SpyTransport';
export type { LogFormatter } from './logger/transports/formatters/LogFormatter';

// --- Interfaces for Instrumented Clients ---
export type { IBeaconRedis } from './redis/IBeaconRedis';
// export type { InstrumentedHttpClient } from './http/types';
