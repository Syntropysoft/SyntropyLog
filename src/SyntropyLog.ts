/**
 * @file src/SyntropyLog.ts
 * @description The main singleton class for the SyntropyLog framework.
 * It orchestrates the initialization and shutdown of all managers.
 */
import { ZodError } from 'zod';
import { SyntropyLogConfig } from './config';
import { syntropyLogConfigSchema } from './config.schema';
import { IContextManager } from './context';
import { ILogger } from './logger';
import { LoggerFactory } from './logger/LoggerFactory';
import { RedisManager } from './redis/RedisManager';
import type { IBeaconRedis } from './redis/IBeaconRedis';
import { sanitizeConfig } from './utils/sanitizeConfig';
import { HttpManager } from './http/HttpManager';
import { InstrumentedHttpClient } from './http/InstrumentedHttpClient';
import { BrokerManager } from './brokers/BrokerManager';
import { InstrumentedBrokerClient } from './brokers/InstrumentedBrokerClient';

/**
 * @class SyntropyLog
 * @description The main entry point and orchestrator for the framework.
 * It follows the Singleton pattern to ensure a single, consistent instance
 * throughout the application's lifecycle.
 */
export class SyntropyLog {
  /** @private The singleton instance. */
  private static instance: SyntropyLog;
  /** @private A flag to ensure the framework is initialized only once. */
  private _isInitialized = false;

  /** @private The parsed and sanitized configuration for the framework. */
  private config!: SyntropyLogConfig;
  /** @private The factory for creating and managing logger instances. */
  private loggerFactory!: LoggerFactory;
  /** @private The manager for Redis client instances. */
  private redisManager!: RedisManager;
  /** @private The manager for HTTP client instances. */
  private httpManager!: HttpManager;
  /** @private The manager for message broker client instances. */
  private brokerManager!: BrokerManager;

  /** @private The constructor is private to enforce the singleton pattern. */
  private constructor() {}

  /**
   * Gets the singleton instance of the SyntropyLog class.
   * @returns {SyntropyLog} The singleton instance.
   */
  public static getInstance(): SyntropyLog {
    if (!SyntropyLog.instance) {
      SyntropyLog.instance = new SyntropyLog();
    }
    return SyntropyLog.instance;
  }

  /**
   * Initializes the framework with the provided configuration.
   * This method sets up all managers and must be called before any other method.
   * @param {SyntropyLogConfig} config - The configuration object for the framework.
   * @returns {Promise<void>}
   */
  public async init(config: SyntropyLogConfig): Promise<void> {
    if (this._isInitialized) {
      console.warn(
        '[SyntropyLog] Warning: Framework has already been initialized. Ignoring subsequent init() call.'
      );
      return;
    }

    try {
      const parsedConfig = syntropyLogConfigSchema.parse(config);
      const sanitizedConfig = sanitizeConfig(parsedConfig);
      this.config = sanitizedConfig; // Store the config for later use (e.g., in shutdown)

      this.loggerFactory = new LoggerFactory(sanitizedConfig);
      const mainLogger = this.loggerFactory.getLogger('syntropylog-main');

      this.redisManager = new RedisManager({
        config: sanitizedConfig.redis,
        logger: this.loggerFactory.getLogger('redis-manager'),
      });

      this.httpManager = new HttpManager({
        config: sanitizedConfig,
        loggerFactory: this.loggerFactory,
        contextManager: this.loggerFactory.getContextManager(),
      });

      this.brokerManager = new BrokerManager({
        config: sanitizedConfig,
        loggerFactory: this.loggerFactory,
        contextManager: this.loggerFactory.getContextManager(),
      });

      this._isInitialized = true;
      mainLogger.info('SyntropyLog framework initialized successfully.');
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(
          '[SyntropyLog] Configuration validation failed:',
          error.errors
        );
      } else {
        console.error('[SyntropyLog] Failed to initialize framework:', error);
      }
      throw error;
    }
  }

  /**
   * Retrieves a logger instance by name.
   * @param {string} [name='default'] - The name of the logger.
   * @returns {ILogger} The logger instance.
   */
  public getLogger(name = 'default'): ILogger {
    this.ensureInitialized();
    return this.loggerFactory.getLogger(name);
  }

  /**
   * Retrieves a managed Redis client instance by name.
   * @param {string} name - The name of the Redis instance.
   * @returns {IBeaconRedis} The Redis client instance.
   */
  public getRedis(name: string): IBeaconRedis {
    this.ensureInitialized();
    return this.redisManager.getInstance(name);
  }

  /**
   * Retrieves a managed and instrumented HTTP client instance by name.
   * @param {string} name - The name of the HTTP client instance.
   * @returns {InstrumentedHttpClient} The HTTP client instance.
   */
  public getHttp(name: string): InstrumentedHttpClient {
    this.ensureInitialized();
    return this.httpManager.getInstance(name);
  }

  /**
   * Retrieves the context manager instance.
   * @returns {IContextManager} The context manager.
   */
  public getContextManager(): IContextManager {
    this.ensureInitialized();
    return this.loggerFactory.getContextManager();
  }

  /**
   * Retrieves a managed and instrumented message broker client instance by name.
   * @param {string} name - The name of the broker client instance.
   * @returns {InstrumentedBrokerClient} The broker client instance.
   */
  public getBroker(name: string): InstrumentedBrokerClient {
    this.ensureInitialized();
    return this.brokerManager.getInstance(name);
  }

  /**
   * Gracefully shuts down all managed clients and flushes log transports.
   * @returns {Promise<void>}
   */
  public async shutdown(): Promise<void> {
    if (!this._isInitialized) return;

    const mainLogger = this.loggerFactory.getLogger('syntropylog-main');
    mainLogger.info('Shutting down SyntropyLog framework...');

    const timeout = this.config.shutdownTimeout ?? 5000;
    const timeoutPromise = new Promise((_, reject) =>
      setTimeout(
        () => reject(new Error(`Shutdown timed out after ${timeout}ms`)),
        timeout
      )
    );

    try {
      const shutdownWork = Promise.allSettled([
        this.redisManager.shutdown(),
        this.httpManager.shutdown(),
        this.brokerManager.shutdown(),
        this.loggerFactory.flushAllTransports(),
      ]);

      await Promise.race([shutdownWork, timeoutPromise]);
      mainLogger.info('SyntropyLog shut down successfully.');
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (error: any) {
      mainLogger.warn('Shutdown process timed out.', {
        detail: 'Some resources may not have been released correctly.',
        error: error.message,
      });
    } finally {
      this._isInitialized = false;
    }
  }

  /**
   * @private
   * Throws an error if the framework has not been initialized.
   */
  private ensureInitialized(): void {
    if (!this._isInitialized) {
      throw new Error(
        'SyntropyLog has not been initialized. Call init() before accessing clients or loggers.'
      );
    }
  }
}

/** The singleton instance of the SyntropyLog framework. */
export const syntropyLog = SyntropyLog.getInstance();
