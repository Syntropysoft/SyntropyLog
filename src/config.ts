/**
 * FILE: src/config.ts
 * DESCRIPTION: Defines and exports the configuration types for the library.
 * These types are inferred directly from the Zod schemas defined in `config.schema.ts`,
 * ensuring that static types and runtime validation are always in sync.
 */

import { z } from 'zod';
import { syntropyLog } from './SyntropyLog';
import {
  syntropyLogConfigSchema, // Using the old name temporarily
  redisInstanceConfigSchema,
  httpInstanceConfigSchema,
  httpConfigSchema,
  redisConfigSchema,
} from './config.schema';

/** The complete, top-level configuration type for the SyntropyLog framework. */
export type SyntropyLogConfig = z.infer<typeof syntropyLogConfigSchema>;

/** The configuration type for a single Redis instance. */
export type RedisInstanceConfig = z.infer<typeof redisInstanceConfigSchema>;

/** The configuration type for a single HTTP client instance. */
export type HttpClientInstanceConfig = z.infer<typeof httpInstanceConfigSchema>;

/**
 * The configuration type for the global HTTP settings.
 * `NonNullable` is used to ensure it's always an object, even if optional in the main config.
 */
export type SyntropyHttpConfig = NonNullable<z.infer<typeof httpConfigSchema>>;

/**
 * The configuration type for the global Redis settings.
 * `NonNullable` is used to ensure it's always an object, even if optional in the main config.
 */
export type SyntropyRedisConfig = NonNullable<z.infer<typeof redisConfigSchema>>;

/**
 * Defines the properties of a Redis instance that can be reconfigured dynamically.
 * Connection-related properties (like url, mode, etc.) are intentionally omitted.
 */
export type RedisInstanceReconfigurableConfig = Pick<
  RedisInstanceConfig,
  'instanceName' | 'logging'
>;
