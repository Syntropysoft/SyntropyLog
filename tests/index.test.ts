/**
 * FILE: tests/index.test.ts
 * DESCRIPTION: Unit tests for the main public API entry point.
 */

import { describe, it, expect } from 'vitest';
import * as allExports from '../src/index';
import { SyntropyLog } from '../src/SyntropyLog';

describe('Main entry point (src/index.ts)', () => {
  it('should export the syntropyLog singleton instance', () => {
    expect(allExports.syntropyLog).toBeDefined();
    // The singleton is an instance of the SyntropyLog class.
    // Importing the class for this check makes the test more robust.
    expect(allExports.syntropyLog).toBeInstanceOf(SyntropyLog);
  });

  it('should export Transport base class and its implementations', () => {
    const transportExports = {
      Transport: allExports.Transport,
      ConsoleTransport: allExports.ConsoleTransport,
      PrettyConsoleTransport: allExports.PrettyConsoleTransport,
      CompactConsoleTransport: allExports.CompactConsoleTransport,
      ClassicConsoleTransport: allExports.ClassicConsoleTransport,
      SpyTransport: allExports.SpyTransport,
    };

    const BaseTransport = allExports.Transport;

    for (const [name, constructor] of Object.entries(transportExports)) {
      expect(constructor, `${name} should be defined`).toBeDefined();
      expect(typeof constructor, `${name} should be a class (function)`).toBe(
        'function'
      );

      // Check that implementations inherit from the base Transport class
      if (name !== 'Transport') {
        // All transport implementations can be instantiated without arguments
        const instance = new constructor();
        expect(instance).toBeInstanceOf(BaseTransport);
      }
    }
  });

  it('should match the public API snapshot to prevent unintentional changes', () => {
    const exportedKeys = Object.keys(allExports).sort();
    expect(exportedKeys).toMatchSnapshot();
    // This ensures that no new exports are added or removed without explicit approval.
    // If you intentionally change the exports, you can update the snapshot.
  });
});