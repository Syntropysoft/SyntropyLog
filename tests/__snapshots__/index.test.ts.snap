// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main entry point (src/index.ts) > should match the public API snapshot to prevent unintentional changes 1`] = `
[
  "ClassicConsoleTransport",
  "CompactConsoleTransport",
  "ConsoleTransport",
  "PrettyConsoleTransport",
  "SpyTransport",
  "Transport",
  "syntropyLog",
]
`;
